### PR TITLE
RPC: add --no-rest-tls option

### PR DESCRIPTION
**Pull request by halseth**
Supersedes #2277, disabling TLS for REST ~~and gRPC~~ connections.

### CODE REVIEW

File: lnd.go
Comment (bhandras) [inline lnd.go:766]: nit: could be just `200 * 1024 * 1024`

File: lnd.go
Comment (bhandras) [inline lnd.go:756]: nb: maybe it's worth considering a struct as a return here given the number of items (or extending the godoc).

**Review by bhandras — COMMENTED** [review]
Looks good, just two comments

File: config.go
Comment (guggero) [inline config.go:207]: We discussed this offline a bit. Just letting the user disable TLS is quite a big safety risk. We don't even let users disable macaroons if they're not at least listening on a private/loopback IP address. Maybe we need to add a condition for TLS in `lncfg.EnforceSafeAuthentication` as well?

File: lnd.go
Comment (guggero) [reply lnd.go:766]: Or if we touch that code anyway, let's make it a constant somewhere that we can also use in `lncli`.

File: lnd.go
Comment (guggero) [inline lnd.go:772]: Don't we need a `grpc.WithInsecure()` here?
Also disabling TLS on gRPC has an impact on `lncli`. We need a `--notls` flag there as well.
And we also need to think about the interactions between this flag and the `--no-macaroons` flag. For example the gRPC client we use for `lncli` will refuse to send the macaroon over a non-TLS connection.

File: config.go
Comment (guggero) [reply config.go:207]: Also the description should say "for RPC and REST connections".

**Review by guggero — COMMENTED** [review]
I like the approach of returning the dial options and listen function instead of the credentials themselves!
I'm not sure about the security implications of this though, maybe we need some more discussion.

File: lnd.go
Comment (NicolasDorier) [inline lnd.go:280]: this makes it quite useless for me. We host LND in a container, the ports are not exposed outside to the host, so we need to bind on 0.0.0.0.

File: config.go
Comment (NicolasDorier) [reply config.go:207]: Please do what @guggero suggest. As explained here https://github.com/lightningnetwork/lnd/pull/4648/files#r495456714 :

1. I have a very valid reason to bind on 0.0.0.0
2. Don't waste time of people silently changing the settings they meant.

File: lnd.go
Comment (guggero) [reply lnd.go:280]: We should add a comment to this block. This only changes the address the REST proxy uses to dial the gRPC server internally. If it would use `0.0.0.0` it would run into the problem of that address not being contained in the TLS certificate (because that will include concrete IP addresses like `127.0.0.1`). The code as it currently is implemented does **not** restrict the gRPC bind address in any way.

Comment (Roasbeef) [reply lnd.go:280]: > this makes it quite useless for me. We host LND in a container, the ports are not exposed outside to the host, we can't predict the network address of the container, so we need to bind on 0.0.0.0.

Depending on the set up, it would be possible to have something like nginx (or any other reverse proxy really) accept/listen on all interfaces, but then forward to the local lnd node (assuming it's in the same "pod").

Comment (NicolasDorier) [reply lnd.go:280]: it is not on same pod. The reverse proxy is in his own container and shared by more than one service.

Comment (halseth) [reply lnd.go:283]: > this makes it quite useless for me. We host LND in a container, the ports are not exposed outside to the host, we can't predict the network address of the container, so we need to bind on 0.0.0.0.

This code is just moved, so no behavior change. Added a comment as @guggero suggested.

File: config.go
Comment (halseth) [reply config.go]: Added `EnforceSafeAuthentication` check for TLS also.

Note that this disallows disabling TLS and macaroons when listening on `0.0.0.0`, as you will have to specify a non-public address to listen on.

File: lnd.go
Comment (halseth) [reply lnd.go]: Made the change REST only.

**Review by guggero — COMMENTED** [review]
We're very close! Just one small bug in the way the REST proxy dials the internal gRPC server.

I was able to bind to the internal docker container IP using the following docker run command:
```
docker run -d \
  -p 8080:8080 \
...
  --restart unless-stopped \
  --name lnd-testnet \
  --hostname lnd-testnet \
...
  guggero/lnd:some-tag \
...
  --restlisten="lnd-testnet:8080" \
  --notls \
...
```

This resulted in the following listening IP:
`[INF] RPCS: password gRPC proxy started at 172.19.0.2:8080`

File: config.go
Comment (guggero) [reply config.go]: Nice! Can confirm this works in a docker setup, see https://github.com/lightningnetwork/lnd/pull/4648#pullrequestreview-499285729.

File: lnd.go
Comment (halseth) [reply lnd.go]: Ah, true, good catch! Fixed 👯

File: lncfg/address.go
Comment (halseth) [reply lncfg/address.go]: Fixed.

**Comment by halseth** [comment]
> We're very close! Just one small bug in the way the REST proxy dials the internal gRPC server.
> 
> I was able to bind to the internal docker container IP using the following docker run command:
> 
> ```
> docker run -d \
>   -p 10009:10009 \
> ### don't expose the REST port to the outside world here as that would be very insecure!!!
> ...
>   --restart unless-stopped \
>   --name lnd-testnet \
>   --hostname lnd-testnet \
> ...
>   guggero/lnd:some-tag \
> ...
>   --restlisten="lnd-testnet:8080" \
>   --notls \
> ...
> ```
> 
> As long as you set the hostname in `--restlilsten` to the same value as you use for the docker flag `--hostname` this will resolve to the internal container IP at runtime.
> 
> Example from my above setup:
> `[INF] RPCS: password gRPC proxy started at 172.19.0.2:8080`
> 
> I was able to reach the container from within the docker network using the hostname as well.

@NicolasDorier can you check if this works for you?

**Review by guggero — APPROVED** [review]
LGTM :taco:

**Comment by halseth** [comment]
Ping @NicolasDorier @rockstardev if this does what you need it is ready for merge.

**Comment by rockstardev** [comment]
If @NicolasDorier says it looks good, I can create a docker container and do a round of testing in our dev env so we're prepped to drop our switch and use yours.

**Comment by NicolasDorier** [comment]
If @rockstardev says it works then that's cool for me!

**Comment by Roasbeef** [comment]
What's this blocked on, if anything?

**Comment by halseth** [comment]
I think we just need an ACK from @rockstardev and @NicolasDorier. 

Does it work for BTCPay?

**Comment by rockstardev** [comment]
@halseth seems to be working - the only tricky problem I ran into was that 0.0.0.0 registration resulted in rest listener working both on HTTP and HTTPS. I've added new variable to address this and we're on path of using this switch of yours.

If you're in crunch mode I say you merge while me and @NicolasDorier do last round of testing. I will still need to keep building custom Docker image anyways until we address other discrepancies, but definitely will be easier. See this PR: 

https://github.com/btcpayserver/BTCPayServer.Lightning/pull/40

File: config.go
Comment (bhandras) [reply config.go:213]: +1

**Review by bhandras — APPROVED** [review]
LGTM :coconut:

**Comment by rockstardev** [comment]
OK, I'm ready to make the switch so BTCPayServer uses this parameter - just let me know when it's merged and released.

**Comment by guggero** [comment]
Needs an addition to `sample-lnd.conf`, but otherwise looks ready to be merged!

**Comment by Roasbeef** [comment]
Any reason why the new arg doesn't specify it's REST only? Also why not also allow for the RPC in general? Just to minimize the diff and only implement what's requested by the user?

**Comment by Roasbeef** [comment]
Travis is backed up, but the config check is failing as Oliver pointed out.

File: config.go
Comment (halseth) [reply config.go]: Done. `no-rest-tls`

**Comment by halseth** [comment]
> Any reason why the new arg doesn't specify it's REST only? Also why not also allow for the RPC in general? Just to minimize the diff and only implement what's requested by the user?

Changed arg name to `no-rest-tls`. Reason we did not add it for RPC in general is to minimize the number of footguns, and also don't have to add the option to `lncli`.

**Comment by 3nprob** [comment]
@halseth Would it be open for consideration to do this for gRPC as well? For scenarios where all TLS of internal and external services are already handled by a service mesh or reverse proxy on the local machine, it's very cubersome to have to engineer a special case just for LND (and arguable more footguns there)

**Comment by halseth** [comment]
> @halseth Would it be open for consideration to do this for gRPC as well? For scenarios where services are running in containers with isolated filesystems and all TLS is already handled by a service mesh or reverse proxy on the local machine, it's very cubersome to have to engineer a special case just for LND (and arguable more footguns there)

It's definitely possible, but we didn't want to enable it for various reasons, most notably as you mention because it's a footgun with seemingly little reward: https://github.com/lightningnetwork/lnd/pull/4648#discussion_r494798090

**Comment by james-callahan** [comment]
> because it's a footgun with seemingly little reward: [#4648 (comment)](https://github.com/lightningnetwork/lnd/pull/4648#discussion_r494798090)

I believe @3nprob was noting the reverse: `lnd` requiring a dance around TLS when nothing else does *is* the footgun.
  - We want to deploy `lnd` behind a reverse proxy that does the TLS termination (which is our standard pattern): there's no way to opt-out of lnd doing the TLS termination.
      - This means we need custom TLS certificate provisioning just for `lnd`
  - We want to do health checks using `lncli` to `localhost`, but that requires the certificate to have `localhost` in it as there is no  (e.g.) `--no-verify` or `--connect-to` (to get it to use one host for SNI/verification but a different one for the actual TCP connection) flag on `lncli`

**Comment by chpio** [comment]
How would one use this with docker networks? I think, you *need to* bind to address 0.0.0.0 to be reachable from another container, but that's not allowed.

**Comment by guggero** [comment]
> How would one use this with docker networks? I think, you _need to_ bind to address 0.0.0.0 to be reachable from another container, but that's not allowed.

By binding to the container host name instead of 0.0.0.0. See example here: https://github.com/lightningnetwork/lnd/pull/4648#pullrequestreview-499285729

**Comment by chpio** [comment]
@guggero Thank you for the answer. My lnd is in two networks, the default network (running bitcoind, tor, etc) and the reverse proxy network. When using the container name it seem to bind to the default network, but i need it to bind to the reverse proxy net. Maybe there's a way to specify the network as well via hostnames, need to investigate further.

**Comment by guggero** [comment]
You can specify multiple listening addresses. So I think it will work as long as the first one is a local network address.

**Comment by chpio** [comment]
So something like
```
restlisten=127.0.0.1:8080
restlisten=0.0.0.0:8080
```

That fails as well

> LTND: Error validating config: ValidateConfig: error enforcing safe authentication on REST ports: detected RPC server listening on publicly reachable interface 0.0.0.0:8080 with encryption disabled! Refusing to start with --no-rest-tls specified

It checks all addresses: https://github.com/halseth/lnd/blob/4a03074fe80e5ac4f4516de1857fcfd9cfb011ea/lncfg/address.go#L52-L79

**Comment by guggero** [comment]
No, I mean something like this:
```
restlisten=127.0.0.1:8080
restlisten=>docker-container-hostname-which-will-resolve-to-a-172-xx-xx-xx-address>:8080

```

**Comment by chpio** [comment]
yeah, that's my problem. How do i get those ip addresses/docker-hostnames. I need to bind to an address from the default network & the reverse-proxy network (to be reachable from the reverse-proxy).

**Comment by guggero** [comment]
I'm not sure, I'm not too familiar with the docker internals. But you could write a shell script, mount that into the container and execute that on container start instead of `lnd` directly. Then within that script you could execute whatever commands you need to figure out the IPs you want to use, then start `lnd` with the appropriate IPs.

I'm sure you're aware of the dangers of not using macaroons, right? Especially since you're planning on exposing the REST API through a reverse proxy. Maybe there is a different way to achieve what you're planning? Can you perhaps elaborate on your use case a bit?

**Comment by chpio** [comment]
Hmm, thank you for the idea, yeah could try making a custom lnd entrypoint/command script.

> I'm sure you're aware of the dangers of not using macaroons, right?

I'm not doing that, i just want the reverse proxy to work, which it can't with self signed tls certs. The proxy is handling the tls connection on the public side anyway.

**Comment by guggero** [comment]
> I'm not doing that, i just want the reverse proxy to work, which it can't with self signed tls certs. The proxy is handling the tls connection on the public side anyway.

Ah, right, was thinking of the wrong flag but same security restriction applies.

**Pull request merged**
